numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -0,0 +1,3063 @@
+    }
+
+    NpyString_release_allocators(3, allocators);
+    return 0;
+
+fail:
+    NpyString_release_allocators(3, allocators);
+    return -1;
+
+}
+
+static NPY_CASTING
+strip_whitespace_resolve_descriptors(
+        struct PyArrayMethodObject_tag *NPY_UNUSED(method),
+        PyArray_DTypeMeta *NPY_UNUSED(dtypes[]),
+        PyArray_Descr *given_descrs[],
+        PyArray_Descr *loop_descrs[],
+        npy_intp *NPY_UNUSED(view_offset))
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    PyArray_Descr *out_descr = NULL;
+
+    if (given_descrs[1] == NULL) {
+        out_descr = (PyArray_Descr *)new_stringdtype_instance(
+                ((PyArray_StringDTypeObject *)given_descrs[0])->na_object,
+                ((PyArray_StringDTypeObject *)given_descrs[0])->coerce);
+
+        if (out_descr == NULL) {
+            return (NPY_CASTING)-1;
+        }
+    }
+    else {
+        Py_INCREF(given_descrs[1]);
+        out_descr = given_descrs[1];
+    }
+
+    loop_descrs[1] = out_descr;
+
+    return NPY_NO_CASTING;
+}
+
+static int
+string_lstrip_whitespace_strided_loop(
+        PyArrayMethod_Context *context,
+        char *const data[], npy_intp const dimensions[],
+        npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_StringDTypeObject *descr = (PyArray_StringDTypeObject *)context->descriptors[0];
+    int has_string_na = descr->has_string_na;
+    int has_null = descr->na_object != NULL;
+    const npy_static_string *default_string = &descr->default_string;
+
+    npy_string_allocator *allocators[2] = {};
+    NpyString_acquire_allocators(2, context->descriptors, allocators);
+    npy_string_allocator *allocator = allocators[0];
+    npy_string_allocator *oallocator = allocators[1];
+
+    char *in = data[0];
+    char *out = data[1];
+
+    npy_intp N = dimensions[0];
+
+    while (N--) {
+        const npy_packed_static_string *ps = (npy_packed_static_string *)in;
+        npy_static_string s = {0, NULL};
+        int s_isnull = NpyString_load(allocator, ps, &s);
+
+        if (s_isnull == -1) {
+            npy_gil_error(PyExc_MemoryError, "Failed to load string in lstrip");
+            goto fail;
+        }
+
+        npy_packed_static_string *ops = (npy_packed_static_string *)out;
+
+        if (NPY_UNLIKELY(s_isnull)) {
+            if (has_string_na || !has_null) {
+                s = *default_string;
+            }
+            else {
+                npy_gil_error(PyExc_TypeError,
+                              "Cannot strip null values that are not strings");
+                goto fail;
+            }
+        }
+
+        char *new_buf = (char *)PyMem_RawCalloc(s.size, 1);
+        Buffer<ENCODING::UTF8> buf((char *)s.buf, s.size);
+        Buffer<ENCODING::UTF8> outbuf(new_buf, s.size);
+        string_lrstrip_whitespace(buf, outbuf, STRIPTYPE::LEFTSTRIP);
+
+        if (NpyString_pack(oallocator, ops, new_buf, strnlen(new_buf, s.size)) < 0) {
```

Comment:
This doesn't look great to me.  Pretty much any use of `strnlen` smells like a bug and I am sure it is here (maybe a niche one for embedded `\0`s, sure).

Can't `string_lstrip_whitespace` just return the length and you then copy things directly?  Otherwise it looks like copying the string twice unnecessarily.

Where we _do_ need an intermediate buffer it would be cool to try to move it out of core loop (growing it if needed), potentially even into the auxdata.
I would expect `malloc` to be significantly slower than the stripping for small strings.

Reply:
(The same comment of course also for rstrip and maybe other stripping, dunno)

Reply:
Yeah, good point, the `strnlen` was definitely a hack. I'll take another look at this.